### PR TITLE
Add two APIs for getting stylistic set labels

### LINKED ISSUE #425: get StyleSet names if available
Scribus uses HarfBuzz to get font features, but right now harfbuzz gives only the tags. This is enough for almost font features expect Stylistic Set 1 - Stylistic Set 20. In [one of scribus bugs](https://bugs.scribus.net/view.php?id=14621), a user requests a feature to display StyleSet names if available instead of just displaying ss01, ss02, ss03, …, ss20. 

Is this possible with current api?

### PR DESCRIPTION
**Pull request by ebraminio**
* hb_ot_layout_get_feature_characters
* hb_ot_layout_get_feature_name_ids

However HarfBuzz currently doesn't expose an API for retrieving the actual
information associated with NameId from the `name` table and that should be
done separately.

Fixes #425 

Uses the test font created by @khaledhosny

### CODE REVIEW

File: src/hb-ot-layout.cc [hunk 1]
```diff
@@ -1116,6 +1116,143 @@ hb_ot_layout_get_size_params (hb_face_t    *face,
...
+				   unsigned int   *num_named_parameters, /* OUT.  May be NULL */
+				   unsigned int   *first_param_id        /* OUT.  May be NULL */)
+{
+  static_assert ((OT::FeatureVariations::NOT_FOUND_INDEX == HB_OT_LAYOUT_NO_VARIATIONS_INDEX), "");
+  const OT::GSUBGPOS &g = get_gsubgpos_table (face, table_tag);
+
+  const OT::Feature &f = g.get_feature_variation (feature_index, variations_index);
+
+  const OT::FeatureParams &feature_params = f.get_feature_params ();
+  if (&feature_params != &Null (OT::FeatureParams))
```
Comment (ebraminio) [inline src/hb-ot-layout.cc]: I know there is a better way for this but I used `size` as the base for this initially which goes for the same way.

File: src/hb-ot-layout.cc [hunk 2]
```diff
@@ -1116,6 +1116,143 @@ hb_ot_layout_get_size_params (hb_face_t    *face,
...
+				   unsigned int   *label_id,             /* OUT.  May be NULL */
+				   unsigned int   *tooltip_id,           /* OUT.  May be NULL */
+				   unsigned int   *sample_id,            /* OUT.  May be NULL */
+				   unsigned int   *num_named_parameters, /* OUT.  May be NULL */
+				   unsigned int   *first_param_id        /* OUT.  May be NULL */)
+{
+  static_assert ((OT::FeatureVariations::NOT_FOUND_INDEX == HB_OT_LAYOUT_NO_VARIATIONS_INDEX), "");
+  const OT::GSUBGPOS &g = get_gsubgpos_table (face, table_tag);
+
+  const OT::Feature &f = g.get_feature_variation (feature_index, variations_index);
```
Comment (ebraminio) [inline src/hb-ot-layout.cc]: Maybe don't care about variation?

File: test/api/test-ot-nameid.c [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    g_error ("Font not found.");
+
+  hb_face_t *face = hb_face_create (blob, 0);
+  hb_font_t *font = hb_font_create (face);
+
+  hb_tag_t cv01 = HB_TAG ('c','v','0','1');
+  unsigned int feature_index = 0;
+  // FIXME: See why below doesn't work
+  // if (!hb_ot_layout_language_find_feature (face, HB_OT_TAG_GSUB, 0, 0, cv01, &feature_index))
+  //   g_error ("Failed to find feature index");
```
Comment (ebraminio) [inline test/api/test-ot-nameid.c]: I liked to know why, maybe later

File: docs/harfbuzz-sections.txt [hunk 4]
```diff
@@ -482,6 +482,8 @@ hb_ot_layout_collect_features
 hb_ot_layout_feature_get_lookups
 hb_ot_layout_feature_with_variations_get_lookups
 hb_ot_layout_get_attach_points
+hb_ot_layout_get_feature_characters
```
Comment (ebraminio) [inline docs/harfbuzz-sections.txt]: Maybe hb_ot_layout_feature_get_characters and hb_ot_layout_feature_get_name_ids now that we are passing feature index also?

File: test/api/test-ot-nameid.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+  unsigned int all_chars;
+  all_chars = hb_ot_layout_get_feature_characters (face, cv01, HB_OT_TAG_GSUB, feature_index,
+						   HB_OT_LAYOUT_NO_VARIATIONS_INDEX,
+						   0, &char_count, characters);
+
+  g_assert (all_chars == 2);
+  g_assert (char_count == 2);
+  g_assert (characters[0] == 10);
+  g_assert (characters[1] == 24030);
```
Comment (ebraminio) [inline test/api/test-ot-nameid.c]: Maybe testing non existing features or pagination also sometime later.

File: docs/harfbuzz-sections.txt (hunk 4 above)
Comment (behdad) [reply docs/harfbuzz-sections.txt]: I think, yes.

File: src/hb-ot-layout.cc (hunk 2 above)
Comment (behdad) [reply src/hb-ot-layout.cc]: I *think* you are correct.

Comment (behdad) [reply src/hb-ot-layout.cc]: Don't know if we need a version without variations.

Upon more thought, for this API, no, I think we should ignore variations.

File: src/hb-ot-layout.h [hunk 6]
```diff
@@ -330,6 +330,30 @@ hb_ot_layout_get_size_params (hb_face_t    *face,
 			      unsigned int *range_start,       /* OUT.  May be NULL */
 			      unsigned int *range_end          /* OUT.  May be NULL */);
 
+#define HB_NO_NAME_ID 0xFFFF
```
Comment (behdad) [inline src/hb-ot-layout.h]: Rename to `HB_NAME_ID_INVALID` please.

Comment (behdad) [reply src/hb-ot-layout.h]: Should we add hb_name_id_t?  Maybe yes.

File: test/api/test-ot-nameid.c [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  if (hb_blob_get_length (blob) == 0)
+    g_error ("Font not found.");
+
+  hb_face_t *face = hb_face_create (blob, 0);
+  hb_font_t *font = hb_font_create (face);
+
+  hb_tag_t cv01 = HB_TAG ('c','v','0','1');
+  unsigned int feature_index = 0;
+  // FIXME: See why below doesn't work
+  // if (!hb_ot_layout_language_find_feature (face, HB_OT_TAG_GSUB, 0, 0, cv01, &feature_index))
```
Comment (behdad) [inline test/api/test-ot-nameid.c]: Humm?  Not very convincing.

Comment (ebraminio) [reply test/api/test-ot-nameid.c]: Is that the correct API even to find that feature_index?

File: src/hb-ot-layout.h [hunk 8]
```diff
@@ -330,6 +330,37 @@ hb_ot_layout_get_size_params (hb_face_t    *face,
...
+
+HB_EXTERN unsigned int
+hb_ot_layout_feature_get_characters (hb_face_t      *face,
+				     hb_tag_t        table_tag,
+				     unsigned int    feature_index,
+				     unsigned int    variations_index,
+				     hb_tag_t        feature_tag,
+				     unsigned int    start_offset,
+				     unsigned int   *char_count    /* IN/OUT.  May be NULL */,
+				     hb_codepoint_t *characters    /* OUT.     May be NULL */);
```
Comment (ebraminio) [inline src/hb-ot-layout.h]: Same here, suggestions welcomed.

File: src/hb-ot-layout.cc (hunk 2 above)
Comment (ebraminio) [reply src/hb-ot-layout.cc]: Variation gone

**Comment by ebraminio** [comment]
PTAL

**Comment by behdad** [comment]
Merged and fixed manually.  Thanks Ebrahim!

**Pull request closed without merging**